[WIP] ENH: transparently read compressed files

norok2 opened the pull request:
Transparently read compressed DICOM files, falling back to uncompressed reading. Support for `gzip` and `bzip2` file compression is expected.

norok2 commented:
The use case here is when you have a compressed DICOM file e.g. s = "myfile.ima.gz".

Before this PR, the following code would not work:
d = dicom.read_file(s)

and something like:
with gzip.open(s, 'rb') as fp: d = dicom.read_file(fp)

is required.

After this PR, the previously not working code now auto-magically recognizes gzip as well as bzip2-compressed DICOM files.

This simple modification is not expected to harm any other portion of the code, especially so because the explicit opening through the correct compression module already works.

darcymason commented:
Hi, 
This is interesting, thank you.  I have a question and couple of suggestions:
* is there a significant performance hit for 'normal' files (not zipped)?  Some users have commented about processing thousands of files at a time.  Adding another function call, a try/except, and an attempted zip open for every file maybe doesn't add a lot of overhead, but I'm not sure how much it would be.  In past profiling, once files were in disk cache, the processing time was dominated by the io.open call.  A modified [time_test.py](https://github.com/pydicom/pydicom/blob/master/tests/performance/time_test.py) could check for this.  If it does add a significant time, can the logic be written the other way -- try open a normal file, if that doesn't work, try gzip/bz2?
* can you add a couple of unit tests?  Just something simple to ensure that a known zip file of each type can be read and some data element in the file returned correctly.  They could be added to `test_filereader.py`, using a zipped version of one of the usual test files.
* the doc string for `read_file()` should mention that the file can be one of these types as well as a normal dicom file

Thanks

scaramallion commented on pydicom/filereader.py:
```diff
@@ -42,6 +42,59 @@
+    --------
+    open(), gzip.open(), bz2.open()
+    """
+    try:
```
Why is the import located here?

norok2 replied:
To save on the import, e.g. if you have a `gz` file, and you load it, there is no need to import `bz2`.

scaramallion replied:
Sure, but if I'm reading hundreds or thousands of files then the import will run every time. What I meant was, why isn't the import at the start of the file?

norok2 replied:
The real import happens only once. So there should not be a significant penalty hit if you happen to have `importlib`, `gzip` and `bz2`, which you should as they are standard modules.
In that scenario, I believe this is mostly a matter of style.
However, you are right in that some non-standard installation, and possibly even Py2.6 could have a performance it for always raising an import error (I am not sure that  `try` is substantially slower than a standard `if`, but this could well be).
If you prefer, I can move imports to the beginning of the file.

scaramallion commented on pydicom/filereader.py:
```diff
@@ -42,6 +42,59 @@
+        import_module = __import__
+
+    zip_module_names = 'gzip', 'bz2'
+    file = None
```
Can you use a variable name that's not a python built-in?

norok2 replied:
Most definitely, I forgot `file` was one of those.

norok2 commented:
Hi,
Thanks for the inputs. I have fixed some of the issues you raised. Unfortunately, the modification now have some (I would say fictitious) conflicts with the `master` branch, and I have no idea on how to "Resolve conflicts" in GitHub.
For the performance, it is not obvious to me where should I modify `tests/performance/time_test.py` to test for the potential penalty introduced by this PR. Changing the order in this approach is not an option, but improvement in performance may be possible if inferring file type from mime type magic or extension, and possibly getting the code verbatim rather than in a function. However, I think this is cleaner.
I have added "unit" tests for the compressed case. Maybe there is something wiser than try to compare two `FileDataset` (the plain and the compressed), as this fails in some cases (which I have blacklisted), but it is not obvious to me where / why they are differing. Also, I think it might be a good idea to refactor tests, there is a bit of unnecessary redundancy in my opinion.
Let me know if you have some suggestion on how to proceed.

norok2 commented:
So, I have been experimenting with this a little more, and I believe that if performance is a real issue, I guess that support for `bz2` could be dropped. This is so because in order to provide the expected functionality for all the range of Python versions that `pydicom` supports, the `bz2` + `python2.6` support basically requires closing and re-opening the file. Maybe a specialized `read_file` could be used, e.g. `read_filez` or something.
Also, I believe the files I have modified follow PEP8 a little more strictly than the current code base. Should this be reverted?

massich commented on pydicom/filereader.py:
```diff
@@ -39,14 +42,17 @@
+
     def __init__(self, fp, stop_when=None, force=False):
-        """Read the preamble and meta info and prepare iterator for remainder of file.
+        """Read the preamble and meta info and prepare iterator for remainder
```
keep the method summary within one line. Then if necessary keep a longer explanation.

"""Parses metadata and sets the iterator

Parses the preamble and meta data and sets the ..
"""

massich commented on pydicom/filereader.py:
```diff
@@ -28,7 +30,8 @@
     stat_available = False
 
 from pydicom.errors import InvalidDicomError
-import pydicom.uid  # for Implicit/Explicit/Little/Big Endian transfer syntax UIDs
```
I would drop the comment all together. But if not, I would do it like this:

```py
# for Implicit/Explicit/Little/Big Endian transfer syntax UIDs
import pydicom.uid
```

massich commented on pydicom/filereader.py:
```diff
@@ -233,9 +240,11 @@ def data_element_generator(fp, is_implicit_VR, is_little_endian,
+            # needed
             # immediately to get the character encoding for other tags
-            if defer_size is not None and length > defer_size and tag != (0x08, 0x05):
+            if defer_size is not None and length > defer_size and tag != (
```
```py
if (defer_size is not None and length > defer_size 
        and tag != (0x08, 0x05)):

```

massich commented on pydicom/filereader.py:
```diff
@@ -248,13 +257,18 @@ def data_element_generator(fp, is_implicit_VR, is_little_endian,
+                                                           bytes2hex(
+                                                               value[:12]),
+                                                           dotdot, value[:12],
+                                                           dotdot))
```
Doesn't PEP8 allow something like this?
```py
logger_debug("%08x: %-34s %s %r %s" % (
            bytes2hex(value[:12]), dotdot, value[:12], dotdot))
```

massich commented on pydicom/filereader.py:
```diff
@@ -816,15 +854,16 @@ def read_dicomdir(filename="DICOMDIR"):
     if not isinstance(ds, DicomDir):
-        msg = u"File '{0}' is not a Media Storage Directory file".format(filename)
+        msg = u"File '{0}' is not a Media Storage Directory file".format(
+            filename)
```
```py
msg = u"File '{0}' is not a Media Storage Directory file"
raise InvalidDicomError(msg.format(filename))
```

massich commented on tests/test_filereader.py:
```diff
@@ -230,68 +269,103 @@ def testRTstruct(self):
         ctfull = read_file(ct_name)
         ctfull_tags = sorted(ctfull.keys())
-        msg = "Tag list of partial CT read (except pixel tag and padding) did not match full read"
+        msg = "Tag list of partial CT read (except pixel tag and padding) did " \
```
??

massich commented on tests/test_filereader.py:
```diff
@@ -328,44 +403,59 @@ def testNestedPrivateSQ(self):
         got = ds.InstanceCreationDate
         expected = "20111130"
-        self.assertEqual(got, expected, "Sample data element after file meta with no group length failed, expected '%s', got '%s'" % (expected, got))
+        self.assertEqual(got, expected,
```
msg = "Sample ...

massich commented on tests/test_filereader.py:
```diff
@@ -422,10 +513,14 @@ def test_long_specific_char_set(self):
-                     b'\x30\x30\x30\x38\x2e\x35\x2e\x31\x2e\x31\x2e\x39\x00\x02\x00\x10\x00' \
-                     b'\x55\x49\x12\x00\x31\x2e\x32\x2e\x38\x34\x30\x2e\x31\x30\x30\x30\x38' \
-                     b'\x2e\x31\x2e\x32\x00\x20\x20\x10\x00\x02\x00\x00\x00\x01\x00\x20\x20' \
+        bytestream = b'\x02\x00\x02\x00\x55\x49\x16\x00\x31\x2e\x32\x2e\x38' \
```
make them all the same size

massich commented on tests/test_filereader.py:
```diff
@@ -454,17 +552,22 @@ def test_no_preamble_command_group_dataset(self):
-                     b'\x02\x00\x02\x00\x55\x49\x16\x00\x31\x2e\x32\x2e\x38\x34\x30\x2e\x31' \
-                     b'\x30\x30\x30\x38\x2e\x35\x2e\x31\x2e\x31\x2e\x39\x00\x02\x00\x10\x00' \
-                     b'\x55\x49\x12\x00\x31\x2e\x32\x2e\x38\x34\x30\x2e\x31\x30\x30\x30\x38' \
+                     b'\x02\x00\x02\x00\x55\x49\x16\x00\x31\x2e\x32\x2e\x38' \
```
idem

massich commented on tests/test_filereader.py:
```diff
@@ -483,9 +586,12 @@ def test_preamble_command_meta_no_dataset(self):
-               b'\x02\x00\x02\x00\x55\x49\x16\x00\x31\x2e\x32\x2e\x38\x34\x30\x2e\x31' \
-               b'\x30\x30\x30\x38\x2e\x35\x2e\x31\x2e\x31\x2e\x39\x00\x02\x00\x10\x00' \
-               b'\x55\x49\x12\x00\x31\x2e\x32\x2e\x38\x34\x30\x2e\x31\x30\x30\x30\x38' \
+               b'\x02\x00\x02\x00\x55\x49\x16\x00\x31\x2e\x32\x2e\x38\x34\x30' \
```
idem

massich commented on tests/test_filereader.py:
```diff
@@ -570,22 +682,24 @@ class ReadDataElementTests(unittest.TestCase):
-
-        self.fp = BytesIO() # Implicit little
+                             b'\x01\x01\x02\x03'  # VR of OL
+        ds.PotentialReasonsForProcedure = ['A', 'B',
```
ds.PotentialReasonsForProcedure = [
           'A', 'B' ....

massich commented:
@norok2 I like the idea of using PEP8, I made some comments. 

I think that we could use this PR as an excuse to discuss some CI issues. Like: Adopting PEP8, and if it is worth to keep supporting 2.6

scaramallion commented:
I'm pretty sure pydicom already aims to follow PEP8 and 257 (using numpy style docstrings) based on #22 and #162, but it's a work in progress. Personally, I'm not too fussed about unit tests strictly following PEP8 since they rarely get refactored and also tend to be easily understandable.

I think dropping official support for python 2.6 (and maybe even 3.3) is a good idea, but as long as there are no major compatibility hassles then I'm happy to support them both.

It might be a good idea to split out the PEP8 changes to a separate PR, but it's up to @darcymason.

A [contributing guide](https://github.com/blog/1184-contributing-guidelines) that lays out what's expected from collaborators would be a nice addition.

norok2 commented:
@massich @scaramallion thanks for the feedback. I'd also separate the two modifications. I just accidentally `autopep`-ed the code in *PyCharm*, and that was the result. It is not something hand-crafted but I guess I could spend 1 hr or so doing it, if there is enough interest.

massich commented:
**autopep** is great, but some times there's more than one way of being complaint with the standard and not all of them are equally readable.

scaramallion commented:
@darcymason are we still interested in this? The PR is massively behind `master` and looking at it again 3 years later it seems like the:

```python
with gzip.open(s, 'rb') as f: 
    d = dcmread(f)
```
pattern is actually probably more pythonic than us trying to fit that functionality into pydicom

darcymason commented:
> ... pattern is actually probably more pythonic than us trying to fit that functionality into pydicom

Yes, I agree, it is likely hard to resolve conflicts now, and there were a couple of questions about performance hits (albeit possibly quite small) for the far more common regular file case.  Since there is an easy workaround as you noted, I'll close this one.
